### PR TITLE
Add NDCG implementation; fixes ##346

### LINKED ISSUE #346: NDCG calculation that produces score identical to the sklearn yet faster
This library and polars framework in general misses some ranking metrics like NDCG. See the implementation in the sklearn https://scikit-learn.org/stable/modules/generated/sklearn.metrics.ndcg_score.html

I have implemented NDCG metric calculation in polars that matches sklearn evaluation results and opened a PR to add it. @abstractqqq would you please have a look? I will post the evaluation results and speed increase later today.

### PR DESCRIPTION
**Pull request by AnastasiiaShukhova**
This PR introduces a high-performance NDCG implementation for Polars, providing a crucial ranking evaluation metric for search and recommendation systems. The implementation matches scikit-learn's NDCG scores while leveraging Polars' efficient data processing capabilities.

## Key Features
- 🚀 **High Performance**: Leverages Polars' native operations for efficient computation
- 🔄 **Tie Handling**: Supports both sequential and averaged handling of tied scores
- 🎯 **Top-K Evaluation**: Configurable cutoff for evaluating top-K items
- 📊 **Group-wise Processing**: Efficiently handles multiple queries/users simultaneously
- ✅ **Verified Accuracy**: Matches scikit-learn's implementation exactly

## Use Cases
- Search engine result ranking evaluation
- Recommendation system performance measurement
- Learning to rank model validation
- A/B testing of ranking algorithms

## Technical Details
- Implements both DCG and NDCG calculations
- Handles edge cases (single-item groups, tied scores)
- Provides flexible configuration through parameters
- Includes comprehensive unit tests

## Example Usage
```python
# Evaluate search results
ndcg = calculate_ndcg(
    search_results,
    relevance_col="relevance",
    score_col="bm25_score",
    group_col="query_id",
    k=5
)

# Evaluate recommendations
ndcg = calculate_ndcg(
    recommendations,
    relevance_col="user_rating",
    score_col="model_score",
    group_col="user_id",
    k=10,
    ignore_ties=True
)
```

## Testing
- Added comprehensive unit tests
- Verified against scikit-learn's implementation
- Tested edge cases and special scenarios
- Included example-based tests

## Performance
The implementation is optimized for Polars' execution engine, providing efficient computation even for large-scale ranking tasks with multiple groups and items.

## Related Issues
Closes #346  (if applicable)

### CODE REVIEW

**Comment by abstractqqq** [comment]
This is a good first step. 

You probably have realized the implementation in the package right now are all expressions, which can be used in both lazy and eager Polars. But your implementation here works only for eager Polars. 

This is the plan I have: 

1. I will merge this PR
2. I will work on these metrics in Rust, which will help us properly use these metrics as expressions. You can also work on it of course.
3. Rewrite some Python side code to turn these functions into expressions. (no need for the df input, can be executed in lazy context)
4. Make it in the next release!

**Review by abstractqqq — APPROVED** [review]

**Comment by AnastasiiaShukhova** [comment]
Hey, thank you so much for prompt reply, I dont know Rust, but I am planning to learn it and contribute to this library more, because it is really cool! So for now go ahead and implement it in Rust, if you have time.

**Comment by abstractqqq** [comment]
> Hey, thank you so much for prompt reply, I dont know Rust, but I am planning to learn it and contribute to this library more, because it is really cool! So for now go ahead and implement it in Rust, if you have time.

I am trying to write this in Rust.. And here are a few things I realized after learning the code, and I want to share with you or anyone else who happens to see this.

1. The `_get_rank_discount_at_k` function can be completely turned into an expression, this means Polars can better optimize this and stay lazy for longer.

```
def _get_rank_discount_at_k(group_col:pl.Expr, k:int | None = None) -> pl.Expr: 
    # This assumes sorted. E.g 
    # df.sort([group_col, sort_by_column], descending=True) has been run
    rk = pl.int_range(1, 1 + pl.len()).over(group_col).alias("rank")
    log2 = math.log(2)
    if k is None:
        return (log2 / rk.log1p()).alias("discount")
    else:
        return ((rk <= abs(k)).cast(pl.Float64) * (log2 / rk.log1p())).alias("discount")
```

2. Let's run more in parallel. E.g the following 3 new columns can be easily put in one with_columns() to be run in parallel. Let's squeeze more performance out :)  And btw the flatten() here seems unnecessary. 

```    
df = df.with_columns((pl.col(relevance_col) * pl.col("discount")).alias("discounted_relevance"))
df = df.with_columns(pl.col('discount').cum_sum().over(group_col).flatten().alias('discount_cumsum'))
df = df.with_columns(
        pl.col(sort_by_column)
        .rank(method='min', descending=True)
        .over(group_col)
        .alias("rank_groups")
    )
```

The overall code is very clear and I am able to rewrite a lot of it in Rust quickly. Thanks for the contribution. You will definitely see this function in the new release coming up.

**Comment by abstractqqq** [comment]
PR: https://github.com/abstractqqq/polars_ds_extension/pull/349

It turns out "almost all" of the dcg/ndcg score can be turned into expressions! (Only a very small portion of the code needs to in Rust)

I also renamed it to `query_ndcg_score` to be more consistent with other names in the package.
